packages/dd-trace/src/check_require_cache.js

```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+/* eslint-disable no-console */
+
+const extractPackageAndModulePath = require('../../datadog-core/src/utils/src/extract-package-and-module-path')
+
+/**
+ * The lowest hanging fruit to debug an app that isn't tracing
+ * properly is to check that it is loaded before any modules
+ * that need to be instrumented. This function checks the
+ * `require.cache` to see if any supported packages have
+ * already been required and prints a warning.
+ *
+ * Note that this only going to work for modules within npm
+ * packages, like `express`, and not internal modules, like
+ * `http`.
+ *
+ * The output isn't necessarily 100% perfect. For example if the
+ * app loads a package we instrument but outside of an
+ * unsupported version then a warning would still be displayed.
+ * This is OK as the tracer should be loaded earlier anyway.
+ */
+module.exports = function () {
+  const packages = require('../../datadog-instrumentations/src/helpers/hooks')
+  const naughties = new Set()
+  let didWarn = false
+
+  for (const pathToModule of Object.keys(require.cache)) {
+    const { pkg } = extractPackageAndModulePath(pathToModule)
+
+    if (naughties.has(pkg)) continue
+    if (!(pkg in packages)) continue
+
+    console.error(`Warning: Package '${pkg}' was loaded before dd-trace! This may break instrumentation.`)
```

Comment:
What happens if tens of thousands of packages were loaded before dd-trace?

Reply:
Also, what about frameworks that load the app from a CLI tool, for example `micro`? These would likely have a lot of warnings that are not possible to fix. Same for CI Visibility probably.

Reply:
> What happens if tens of thousands of packages were loaded before dd-trace?

It will print at most 1 message for each package name. E.g. if it finds 10 redis entries there will be one line printed.

> Also, what about frameworks that load the app from a CLI tool, for example micro? These would likely have a lot of warnings that are not possible to fix. Same for CI Visibility probably.

These logs only appear when in debug mode wherein a user should be expecting lots of debug logs.